File: Payload Computer/wp_trigger.py
#!/usr/bin/env python

"""
This code will trigger when a WP is reached and when the linear velocity is below vel_threshold.
It will trigger a flag which in turn will begin the cal sequence at each WP which in turn will 
begin saving metadata. When sequence is completed, another flag will trigger on write_WPs.py 
which will update the WP table. 
If the sequence does not complete within a specified time, the WP table will be updated anyway.

author: Krishna Makhija
rev: 25th April 2021
"""

import rospy, time
import numpy as np
from threading import Event
from mavros_msgs.msg import *
from mavros_msgs.srv import *
from std_msgs.msg import Float32
from geometry_msgs.msg import PoseStamped
from mavros_msgs.msg import WaypointReached, WaypointList, PositionTarget
from sensor_msgs.msg import Imu, NavSatFix
from nav_msgs.msg import Odometry
from tf.transformations import quaternion_from_euler, euler_from_quaternion


n               = 1
event           = Event()
vel_threshold   = 0.2      # linear vel threshold below which drone is considered "stationary" (m/s)
wp_num          = 1
#rospy.set_param('trigger/waypoint', False)
rospy.set_param('trigger/sequence', False)
seq_timeout     = 30
error_tolerance = 1.0       ## distance in m from where to begin sequence
orien_tolerance = 10.0      ## how much pitch and roll per wp is allowed (deg)
wp_wait_timeout = 10


def get_velocity(data):
    """
    Get the current magnitude of linear velocity of the drone.
    """
    global v
    global roll, pitch
    x_vel = data.twist.twist.linear.x
    y_vel = data.twist.twist.linear.y
    z_vel = data.twist.twist.linear.z
    v = (x_vel**2 + y_vel**2 + z_vel**2)**0.5       # magnitude of linear velocity
    q = [data.pose.pose.orientation.x, data.pose.pose.orientation.y,    # angular orientation
                data.pose.pose.orientation.z, data.pose.pose.orientation.w] # in quarternion form
    euler_angs = np.rad2deg(euler_from_quaternion(q, axes='sxyz'))
    roll = euler_angs[0]
    pitch = euler_angs[1]


def wp_reached(data):
    """
    Set trigger when arrived at WP.
    """
    global n
    global sequence
    global wp_num
    if n == 1:
        sequence = data.header.seq
        n = 2
        print("The sequence value is set at %s" %sequence)
    print("The current drone sequence on the FCU is %s" %data.header.seq)
    if data.header.seq == sequence + 1:
        print("Begin countdown to updating WP table: %s seconds." %seq_timeout)
        timeout_event.set()
    if data.header.seq == sequence + 2:        
        print("WP reached: %s. Waiting for drone to be stationary." %wp_num)
        wp_num = wp_num + 1
        event.set()
        timeout_event.clear()
        sequence = data.header.seq


def haversine(lat1, long1, lat2, long2):
    """
    Calculate distance between two points given lat/long coordinates.
    REFERENCE: https://www.movable-type.co.uk/scripts/latlong.html
    """
    r = 6.3781e6                # radius of earth
    phi1      = np.deg2rad(lat1)
    phi2      = np.deg2rad(lat2)
    lam1      = np.deg2rad(long1)
    lam2      = np.deg2rad(long2)
    delta_phi = phi2 - phi1
    delta_lam = lam2 - lam1
    a         = np.sin(delta_phi/2) * np.sin(delta_phi/2) + np.cos(lat1) * np.cos(lat2) * np.sin(delta_lam/2) * np.sin(delta_lam/2)
    c         = 2 * np.arctan2(np.sqrt(a), np.sqrt(1 - a))
    d         = r * c           # distance
    return d


def haversine_3d(lat1, long1, alt1, lat2, long2, alt2):
    """
    Calculate haversine distance in 3 dimensions i.e. Euclidean distance using lat / long coordinates
    """
    alt_diff = (alt2 - alt1)
    d_3d = ((haversine(lat1, long1, lat2, long2))**2 + alt_diff**2)**0.5
    return d_3d


def get_waypoints(data):
    """
    Look up waypoints in FC and "target" them.
    """
    global wp_x_lat
    global wp_y_long
    global wp_z_alt
    try:
        wp_list = data.waypoints
        # skip first two waypoints, i.e. home and takeoff
        target_wp = wp_list[2:]
        wp_x_lat = target_wp[0].x_lat
        wp_y_long = target_wp[0].y_long
        wp_z_alt = target_wp[0].z_alt
        print("Retrieved WP list.")
        print("The current target WP coords are: %s, %s, and %s" %(wp_x_lat, wp_y_long, wp_z_alt))
    except IndexError:
        pass


def get_haversine(data):
    """
    Calculate 2D haversine distance to target using real-time GPS data
    """
    global h
    #while True:
    #time.sleep(0.01)
    if data.status.status == 0:
        if event.is_set() == False:
            h = haversine(data.latitude, data.longitude, wp_x_lat, wp_y_long)
        event.set()
    elif data.status.status == -1:
        print('GPS fix not available.')


def get_distance(data):
    """
    Calculate 3D haversine distance to target.
    This will also begin the entire sequence.
    """
    global distance
    if event.is_set():
        try:
            alt_diff = wp_z_alt - data.pose.position.z
            distance = (h**2 + alt_diff**2)**0.5
            #print('The closest WP is: %s m away.' %(distance))
            event.clear()
            ## * check to see if drone is stable enough to start calibration
            if distance <= error_tolerance and v <= vel_threshold and (roll <= orien_tolerance and pitch <= orien_tolerance):
                print(">>>>WP reached<<< ||| Drone is stable and (almost) not moving.")
                #rospy.set_param('trigger/waypoint', True)
                time.sleep(2)
                rospy.set_param('trigger/sequence', True)
                #FIXME: this is another open loop. what do? can't seem to avoid them
                time.sleep(wp_wait_timeout)
        except IndexError:
            print("index error")
            pass
        except NameError:
            print("Waypoints not received from FCU.")
            pass


def main():
    global get_mission
    try:
        rospy.init_node('wp_trigger', anonymous = True)
        rospy.Subscriber('/mavros/mission/waypoints', WaypointList, get_waypoints)
        rospy.Subscriber('/mavros/global_position/global', NavSatFix, get_haversine)
        rospy.Subscriber('/mavros/local_position/pose', PoseStamped, get_distance)
        rospy.Subscriber('/mavros/local_position/odom', Odometry, get_velocity)
        rospy.spin()
    except (rospy.ROSInterruptException):
        pass


def main_seq():
    try:
        rospy.init_node('wp_trigger', anonymous = True)
        rospy.Subscriber('/mavros/mission/reached', WaypointReached, wp_reached)
        rospy.Subscriber('/mavros/local_position/odom', Odometry, get_velocity)
        rospy.spin()
    except (rospy.ROSInterruptException):
        pass


if __name__ == '__main__':
    main()

File: src/wp_trigger.cpp
/*
* C++ version of wp_trigger.py
* This runs in the background to check if drone is within specc'ed distance of WP and stable 
* enough prior to calibration. It begins the entire sequence by setting the ("trigger/sequence")
* flag.

* author: KM
* date: 28th Aug 2021
*/ 

#include "ros/ros.h"
#include "sensor_msgs/Imu.h"
#include "nav_msgs/Odometry.h"
#include "geometry_msgs/PoseStamped.h"
#include "sensor_msgs/NavSatFix.h"
#include "mavros_msgs/WaypointList.h"
#include "mavros_msgs/Waypoint.h"
#include <iostream>
#include <math.h>
#include <cmath>
#include <thread>
#include <mutex>
#include <unistd.h>

double vel_threshold    = 0.35;     // linear velocity threshold for starting cal (m/s)
double pos_error_tol    = 1.0;      // acceptable 3D radius around setpoint to start cal (m)
double wp_wait_timeout  = 10.0;     // sleep code for this much time after flag is set (s)
double h, v;                        // global 2D haversince distance and linear velocity (m, m/s)
std::vector<mavros_msgs::Waypoint> wp_list;
std::mutex h_mtx;                   // haversine calculation event

// Calculate 2D haversine distance between two points.
double haversine(double lat1, double long1, double lat2, double long2){
    double r = 6.3781e6;            // radius of earth (m)
    double phi1 = lat1 * M_PI / 180;
    double phi2 = lat2 * M_PI / 180;
    double lam1 = long1 * M_PI / 180;
    double lam2 = long2 * M_PI / 180;
    double delta_phi = phi2 - phi1;
    double delta_lam = lam2 - lam1;
    double a = sin(delta_phi/2)*sin(delta_phi/2) + cos(lat1)*cos(lat2)*sin(delta_lam/2)*sin(delta_lam/2);
    double c = 2 * atan2(sqrt(a), sqrt(1 - a));
    double d = r * c;
    return d;
}

// Calculate 3D haversine distance between two points.
double haversine_3d(double lat1, double long1, double alt1, double lat2, double long2, double alt2){
    double alt_diff = alt2 - alt1;
    double d_2d = haversine(lat1, long1, lat2, long2);
    double d_3d = sqrt(pow(d_2d, 2) + pow(alt_diff, 2));
    return d_3d;
}

// Retrieve linear velocities in x, y, and z, and calculate the magnitude of all three directions
void get_velocity(const nav_msgs::Odometry::ConstPtr& msg){
    double x_vel = msg->twist.twist.linear.x;
    double y_vel = msg->twist.twist.linear.y;
    double z_vel = msg->twist.twist.linear.z;
    double v = sqrt(pow(x_vel, 2) + pow(y_vel, 2) + pow(z_vel, 2));
    //ROS_INFO("v = %f", v);
}

// Retrieve current waypoints on the FCU
//! make sure the indices are correct here
void get_waypoints(const mavros_msgs::WaypointList& msg){
    wp_list = msg.waypoints;
    ROS_INFO("Retrieved WP list");
    ROS_INFO("The current target WP coords are %f, %f, and %f", wp_list[1].x_lat, wp_list[1].y_long, wp_list[1].z_alt);
}

// Calculate in real-time what the 2D distance is to the current WP
void get_haversine(const sensor_msgs::NavSatFix::ConstPtr& msg){
    if (msg->status.status == 0){
        h = haversine(msg->latitude, msg->longitude, wp_list[1].x_lat, wp_list[1].y_long);
        //ROS_INFO("Distance to next WP is %f m", h);
        h_mtx.lock();   // lock this thread so that 3D distance can be computed
    }
    else if (msg->status.status == -1){
        ROS_ERROR("GPS signal not available.");
    }
}

// Calculate in real-time what teh 3D distance is to the current WP
void get_distance(const geometry_msgs::PoseStamped::ConstPtr& msg){
    if (!h_mtx.try_lock()){     // ensure that 2D distance is first calculated
        double alt_diff = wp_list[1].z_alt - msg->pose.position.z;
        double distance = sqrt(pow(distance, 2) + pow(h, 2));
        // check to see if drone stability conditions are met
        if (distance < pos_error_tol && v <= vel_threshold){
            ROS_INFO(">>>>WP reached<<< ||| Drone is stable and (almost) not moving.");
            // this begins the whole sequence. will trigger/metadata next.
            ros::param::set("trigger/sequence", true);
            sleep(wp_wait_timeout); // ! can I have the sleep be interrupted by get_waypoints()?
        }
        h_mtx.unlock();         // release lock so that 2D distance can be calculated
    }
}

int main(int argc, char **argv){
    ros::init(argc, argv, "wp_trigger");
    ros::NodeHandle n;
    n.setParam("trigger/sequence", false);
    ros::Subscriber sub1 = n.subscribe("/mavros/local_position/odom", 1000, get_velocity);
    ros::Subscriber sub2 = n.subscribe("/mavros/mission/waypoints", 1000, get_waypoints);
    ros::Subscriber sub3 = n.subscribe("/mavros/global_position/global", 1000, get_haversine);
    ros::Subscriber sub4 = n.subscribe("/mavros/local_position/pose", 1000, get_distance);
    ros::spin();
    return 0;
}
